languages/csharp/AppEncryption/AppEncryption/SessionFactory.cs

```diff
@@ -128,18 +157,173 @@ internal Partition GetPartition(string partitionId)
+        /// Atomically marks a shared <code>CachedSession</code> in the session cache as no longer being used by the
+        /// current caller for the <code>partitionId</code>. This is used to track the number of concurrent users so
+        /// cache eviction policies don't remove an entry while it's still potentially in use.
+        /// </summary>
+        ///
+        /// <param name="partitionId">the partition id for a session</param>
+        private void ReleaseShared(string partitionId)
+        {
+            try
+            {
+                CachedSession cacheItem = SessionCache.Get<CachedSession>(partitionId);
+
+                // Decrements the usage counter of the entry
+                cacheItem.DecrementUsageTracker();
+
+                // If we know it's still in use, don't expire it yet
+                if (!cacheItem.IsUsed())
+                {
+                    // No longer in use, so now kickoff the expire timer
+                    var cacheEntryOptions = new MemoryCacheEntryOptions()
+                        .SetPriority(CacheItemPriority.Low)
+                        .SetSlidingExpiration(TimeSpan.FromMilliseconds(cryptoPolicy.GetSessionCacheExpireMillis()))
+                        .RegisterPostEvictionCallback((id, value, reason, state) =>
+                        {
+                            ((CachedSession)value).GetEnvelopeEncryptionJsonImpl().Dispose();
+                        });
+                    SessionCache.Set(partitionId, cacheItem, cacheEntryOptions);
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.LogError(e, "Unexpected exception during dispose");
+            }
+        }
+
+        private IEnvelopeEncryption<JObject> GetEnvelopeEncryptionJson(string partitionId)
         {
-            Partition partition = GetPartition(partitionId);
+            // Wrap the creation logic in a lambda so the cache entry acquisition can create a new instance when needed
+            Func<CachedSession> createSessionFunc = () =>
+            {
+                Partition partition = GetPartition(partitionId);
 
-            return new EnvelopeEncryptionJsonImpl(
-                partition,
-                metastore,
-                systemKeyCache,
-                secureCryptoKeyDictionaryFactory,
-                new BouncyAes256GcmCrypto(),
-                cryptoPolicy,
-                keyManagementService);
+                EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl = new EnvelopeEncryptionJsonImpl(
+                    partition,
+                    metastore,
+                    systemKeyCache,
+                    new SecureCryptoKeyDictionary<DateTimeOffset>(cryptoPolicy.GetRevokeCheckPeriodMillis()),
+                    new BouncyAes256GcmCrypto(),
+                    cryptoPolicy,
+                    keyManagementService);
+
+                return new CachedSession(envelopeEncryptionJsonImpl, partitionId, this);
```

Comment:
this was because we needed the `partitionId` and an instance of `this` in the inner class. discussed briefly about this offline, can clarify further if required.

Reply:
Code changed to be consistent with Java